MonoGame.Framework/Graphics/Vertices/VertexBuffer.DirectX.cs

```diff
@@ -138,18 +138,52 @@ void GenerateIfRequired()
                 var startBytes = startIndex * elementSizeInBytes;
                 var dataPtr = (IntPtr)(dataHandle.AddrOfPinnedObject().ToInt64() + startBytes);
 
-                var box = new SharpDX.DataBox(dataPtr, 1, 0);
+                var d3dContext = GraphicsDevice._d3dContext;
+                lock (d3dContext)
+                {
+                    if (vertexStride == elementSizeInBytes)
+                    {
+                        var box = new SharpDX.DataBox(dataPtr, 1, 0);
 
-                var region = new SharpDX.Direct3D11.ResourceRegion();
-                region.Top = 0;
-                region.Front = 0;
-                region.Back = 1;
-                region.Bottom = 1;
-                region.Left = offsetInBytes;
-                region.Right = offsetInBytes + (elementCount * elementSizeInBytes);
+                        var region = new SharpDX.Direct3D11.ResourceRegion();
+                        region.Top = 0;
+                        region.Front = 0;
+                        region.Back = 1;
+                        region.Bottom = 1;
+                        region.Left = offsetInBytes;
+                        region.Right = offsetInBytes + (elementCount * elementSizeInBytes);
 
-                lock (GraphicsDevice._d3dContext)
-                    GraphicsDevice._d3dContext.UpdateSubresource(box, _buffer, 0, region);
+                        d3dContext.UpdateSubresource(box, _buffer, 0, region);
+                    }
+                    else
+                    {
+                        // Copy the buffer to a staging resource, so that any elements we don't write to will still be correct.
+                        var stagingDesc = _buffer.Description;
+                        stagingDesc.BindFlags = SharpDX.Direct3D11.BindFlags.None;
+                        stagingDesc.CpuAccessFlags = SharpDX.Direct3D11.CpuAccessFlags.Read | SharpDX.Direct3D11.CpuAccessFlags.Write;
+                        stagingDesc.Usage = SharpDX.Direct3D11.ResourceUsage.Staging;
+                        stagingDesc.OptionFlags = SharpDX.Direct3D11.ResourceOptionFlags.None;
+                        var stagingBuffer = new SharpDX.Direct3D11.Buffer(GraphicsDevice._d3dDevice, stagingDesc);
+
+                        lock (GraphicsDevice._d3dContext)
```

Comment:
That's my bad (although nested locks on the same object don't have any effect, but it's still untidy). I'll fix it.